File: client/src/main/scala/org/totalgrid/reef/client/factory/ReefReconnectingFactory.scala
package org.totalgrid.reef.client.factory

/**
 * Copyright 2011 Green Energy Corp.
 *
 * Licensed to Green Energy Corp (www.greenenergycorp.com) under one or more
 * contributor license agreements. See the NOTICE file distributed with this
 * work for additional information regarding copyright ownership. Green Energy
 * Corp licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
import org.totalgrid.reef.client.settings.AmqpSettings
import org.totalgrid.reef.client.sapi.client.rest.{ ConnectionWatcher => SConnectionWatcher }
import org.totalgrid.reef.client.javaimpl.ConnectionWrapper
import org.totalgrid.reef.broker.qpid.QpidBrokerConnectionFactory
import net.agileautomata.executor4s.Executors
import org.totalgrid.reef.broker.BrokerConnection
import org.totalgrid.reef.client.sapi.client.rest.impl.{ DefaultConnection, DefaultReconnectingFactory }
import org.totalgrid.reef.client.{ ServicesList, ReconnectingConnectionFactory, ConnectionWatcher }
import org.totalgrid.reef.client.ServicesList

class ReefReconnectingFactory(amqpSettings: AmqpSettings, servicesList: ServicesList, startDelay: Long, maxDelay: Long)
    extends ReconnectingConnectionFactory with SConnectionWatcher {

  private val brokerFactory = new QpidBrokerConnectionFactory(amqpSettings)
  private val exe = Executors.newScheduledThreadPool(5)

  private val factory = new DefaultReconnectingFactory(brokerFactory, exe, startDelay, maxDelay)
  factory.addConnectionWatcher(this)

  private var watchers = Set.empty[ConnectionWatcher]

  def addConnectionWatcher(watcher: ConnectionWatcher) = this.synchronized { watchers += watcher }
  def removeConnectionWatcher(watcher: ConnectionWatcher) = this.synchronized { watchers -= watcher }

  def onConnectionClosed(expected: Boolean) =
    this.synchronized { watchers.foreach { _.onConnectionClosed(expected) } }

  def onConnectionOpened(broker: BrokerConnection) = {
    val connection = new ConnectionWrapper(new DefaultConnection(broker, exe, 5000))
    connection.addServicesList(servicesList)
    this.synchronized { watchers.foreach { _.onConnectionOpened(connection) } }
  }

  def start() = factory.start()
  def stop() = {
    factory.stop()
    exe.terminate()
  }
}

File: client/src/main/java/org/totalgrid/reef/client/factory/ReefReconnectingFactory.java
/**
 * Copyright 2011 Green Energy Corp.
 * 
 * Licensed to Green Energy Corp (www.greenenergycorp.com) under one or more
 * contributor license agreements. See the NOTICE file distributed with this
 * work for additional information regarding copyright ownership. Green Energy
 * Corp licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.totalgrid.reef.client.factory;

import net.agileautomata.executor4s.ExecutorService;
import net.agileautomata.executor4s.Executors;
import net.agileautomata.executor4s.Minutes;
import org.totalgrid.reef.broker.BrokerConnection;
import org.totalgrid.reef.broker.BrokerConnectionFactory;
import org.totalgrid.reef.broker.qpid.QpidBrokerConnectionFactory;
import org.totalgrid.reef.client.Connection;
import org.totalgrid.reef.client.ConnectionWatcher;
import org.totalgrid.reef.client.ReconnectingConnectionFactory;
import org.totalgrid.reef.client.ServicesList;
import org.totalgrid.reef.client.javaimpl.ConnectionWrapper;
import org.totalgrid.reef.client.sapi.client.rest.impl.DefaultConnection;
import org.totalgrid.reef.client.sapi.client.rest.impl.DefaultReconnectingFactory;
import org.totalgrid.reef.client.settings.AmqpSettings;

import java.util.HashSet;
import java.util.Set;

/**
 * implementation of reconnecting factory for java applications.
 */
public class ReefReconnectingFactory implements ReconnectingConnectionFactory, org.totalgrid.reef.client.sapi.client.rest.ConnectionWatcher
{

    private Set<ConnectionWatcher> watchers = new HashSet<ConnectionWatcher>();
    private final BrokerConnectionFactory brokerConnectionFactory;
    private final ExecutorService exe;
    private final ServicesList servicesList;

    private final DefaultReconnectingFactory factory;

    /**
     * @param settings amqp settings
     * @param list services list from service-client package
     * @param startDelayMs beginning delay if can't connect first time
     * @param maxDelayMs delay doubles in length upto this maxTime
     */
    public ReefReconnectingFactory( AmqpSettings settings, ServicesList list, long startDelayMs, long maxDelayMs )
    {
        brokerConnectionFactory = new QpidBrokerConnectionFactory( settings );
        exe = Executors.newResizingThreadPool( new Minutes( 5 ) );
        servicesList = list;
        factory = new DefaultReconnectingFactory( brokerConnectionFactory, exe, startDelayMs, maxDelayMs );
        factory.addConnectionWatcher( this );
    }

    public synchronized void onConnectionClosed( boolean expected )
    {
        for ( ConnectionWatcher cw : watchers )
        {
            cw.onConnectionClosed( expected );
        }
    }

    public synchronized void onConnectionOpened( BrokerConnection connection )
    {
        org.totalgrid.reef.client.sapi.client.rest.impl.DefaultConnection scalaConnection;
        scalaConnection = new DefaultConnection( connection, exe, 5000 );
        scalaConnection.addServicesList( servicesList );
        Connection c = new ConnectionWrapper( scalaConnection, exe );
        for ( ConnectionWatcher cw : watchers )
        {
            cw.onConnectionOpened( c );
        }
    }

    public synchronized void addConnectionWatcher( ConnectionWatcher watcher )
    {
        watchers.add( watcher );
    }

    public synchronized void removeConnectionWatcher( ConnectionWatcher watcher )
    {
        watchers.remove( watcher );
    }

    public void start()
    {
        factory.start();
    }

    public void stop()
    {
        factory.stop();
        exe.terminate();
    }
}
